dns-resolver-operator/src/charm.py

```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env python3
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Charm for bind."""
+
+import logging
+import typing
+
+import ops
+from charms.dns_authority.v0 import dns_authority
+
+from bind import BindService
+
+logger = logging.getLogger(__name__)
+
+
+class DnsResolverCharm(ops.CharmBase):
+    """Charm the service."""
+
+    def __init__(self, *args: typing.Any):
+        """Construct.
+
+        Args:
+            args: Arguments passed to the CharmBase parent constructor.
+        """
+        super().__init__(*args)
+        self.bind = BindService()
+        self.dns_authority = dns_authority.DNSAuthorityRequires(self)
+
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.start, self._on_start)
+        self.framework.observe(self.on.stop, self._on_stop)
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
+        self.framework.observe(self.on.collect_unit_status, self._on_collect_status)
+        self.framework.observe(
+            self.on.dns_authority_relation_joined, self._on_dns_authority_relation_joined
+        )
+        self.framework.observe(
+            self.on.dns_authority_relation_changed, self._on_dns_authority_relation_changed
+        )
+        self.unit.open_port("tcp", 53)  # Bind DNS
+        self.unit.open_port("udp", 53)  # Bind DNS
+
+    def _on_collect_status(self, event: ops.CollectStatusEvent) -> None:
+        """Handle collect status event.
+
+        Args:
+            event: Event triggering the collect-status hook
+        """
+        if not self.dns_authority.is_related():
+            event.add_status(ops.BlockedStatus("Needs to be related with an authority charm"))
+            return
+        relation_data = self.dns_authority.get_relation_data()
+        if not relation_data:
+            event.add_status(ops.WaitingStatus("DNS authority relation is not ready"))
+            return
+        if not relation_data.addresses or not relation_data.zones:
+            event.add_status(ops.WaitingStatus("DNS authority relation is empty"))
+            return
+        event.add_status(ops.ActiveStatus())
+
+    def _reconcile(self) -> None:
+        """Reconcile loop."""
+        data = self.dns_authority.get_relation_data()
+        if data is not None:
+            self.bind.update_config_and_reload(data.zones, [str(a) for a in data.addresses])
+        else:
+            self.bind.update_config_and_reload()
+
+    def _on_dns_authority_relation_joined(self, _: ops.RelationJoinedEvent) -> None:
```

Comment:
In my last Synapse review with Grég, he mentioned that we can relax the rule for having a method for each event. We can change the self.framework.observe to call _reconcile directly

Reply:
Changed